### PR TITLE
Adding a mingw based build toolchain for windows

### PR DESCRIPTION
**Pull request by jaym**
This toolchain is similar to the one used in Rubyinstaller's
DevKit. It is slightly different in that it uses tdb-gcc so
that it can build both 32-bit and 64-bit binaries. It should
be almost identical to the things ksubrama did in omnibus.

There are some idempotence issues with this implementation:
* Once a package has been downloaded, it won't be unpacked
  again after the first try. This means if the first go
  fails, the next chef run won't fix itself

### CODE REVIEW

**Comment by jaym** [comment]
Also, build failures probably aren't because of this, master is failing. I can't even run rspec tests locally, it trys to upload my workspace directory to hosted chef I think

**Comment by jaym** [comment]
Test failures should be fixed by https://github.com/chef-cookbooks/build-essential/pull/81

File: recipes/_windows.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+require 'ostruct'
+
+include_recipe '7-zip::default'
+
+def p(url, checksum)
+  OpenStruct.new(url: url, checksum: checksum)
+end
```
Comment (schisamo) [inline recipes/_windows.rb]: This should be in a library. I would create `libraries/_helper.rb` and use https://github.com/chef-cookbooks/languages/blob/master/libraries/_helper.rb as a model.

**Comment by jaym** [comment]
@schisamo moved the function into a library

**Comment by jaym** [comment]
cc @chef-cookbooks/engineering-services

**Comment by scotthain** [comment]
:+1:

**Comment by yzl** [comment]
:+1:

File: attributes/default.rb [hunk 2]
```diff
@@ -18,3 +18,4 @@
 #
 
 default['build-essential']['compile_time'] = false
+default['build-essential']['msys']['path'] = 'C:/msys'
```
Comment (schisamo) [inline attributes/default.rb]: So following the pattern we use for other Windows cookbooks like `wix` and `7-zip` I would make this:

```ruby
default['build-essential']['msys']['path'] = "#{ENV['SYSTEMDRIVE']}\\msys"
```
or
```ruby
default['build-essential']['msys']['path'] = File.join(ENV['SYSTEMDRIVE'], 'msys')
```

**Comment by jaym** [comment]
@schisamo attribute now defaults to `SYSTEMDRIVE/mingw`

**Comment by schisamo** [comment]
:+1:

**Comment by kernelsmith** [comment]
unless I'm using this wrong, this immediately blows up on non windows systems as SYSTEMDRIVE env var does not exist so File.join fails due to inability to implicitly convert from nil to string.

**Comment by jaym** [comment]
this is a good point

**Comment by kernelsmith** [comment]
also, the readme is now out of date https://github.com/chef-cookbooks/build-essential#attributes as this PR added `default['build-essential']['msys']['path']`.  I would pull request that, but given my issues w/the PR as it stands, I figure I should hold off

**Comment by jaym** [comment]
https://github.com/chef-cookbooks/build-essential/pull/83

**Comment by kernelsmith** [comment]
thanks @jaym

**Pull request merged**